src/librustc_mir/interpret/operand.rs

```diff
@@ -536,7 +536,7 @@ impl<'mir, 'tcx, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
                 // potentially requiring the current static to be evaluated again. This is not a
                 // problem here, because we are building an operand which means an actual read is
                 // happening.
-                return Ok(OpTy::from(self.const_eval(GlobalId { instance, promoted })?));
+                return Ok(self.const_eval(GlobalId { instance, promoted }, val.ty)?);
```

Comment:
Curiously, this is now using the use-site type instead of the def-site type, which leads to changing the way we ICE in https://github.com/rust-lang/rust/issues/70942 (we should be ICEing earlier, but after this PR we only ICE very late).

Reply:
The bug is argably unsound `Instance::resolve` behavior, as the mistyped associated `const` in the `impl` is resolved, despite being the wrong type.

Reply:
Hmm, in an `Unevaluated(def_id, substs, None)`, `tcx.type_of(def_id).subst(tcx, substs)` should match the type of the constant, right? Maybe if `Instance::resolve` returns a different `DefId` (i.e. of an associated `const` in an `impl`), we can take the type *from there*?

I'd say "we should require the type is the same between `trait` and `impl`" but I'm worried that is too strict, given how the type-checking requirement works (it's using an inference context).

Oh, but unlike an associated `fn`, an associated *`const`* can't have its own generics, so maybe there's no case in which the types would be equal.

We can use `delay_span_bug` + returning unnormalized constant (or even `tcx.consts.err`).

Reply:
I keep forgetting things to throw in here, oops! I meant to say that I will try to prepare a patch.

Reply:
I was about to ask for clarification, but I think a patch is indeed the best kind of clarification. ;)

Reply:
Ended up doing it in `Instance::resolve` outright: #70970.